controllers/automation/policyautomation_controller.go

```diff
@@ -89,6 +89,100 @@ func (r *PolicyAutomationReconciler) setOwnerReferences(
 	return nil
 }
 
+// getTargetListMap will convert slice targetList to map for search efficiency
+func getTargetListMap(targetList []string) map[string]bool {
+	targetListMap := map[string]bool{}
+	for _, target := range targetList {
+		targetListMap[target] = true
+	}
+
+	return targetListMap
+}
+
+// getViolationContext will put the root policy information into violationContext
+// It also put the status of the non-compliance replicated policies into violationContext
+func (r *PolicyAutomationReconciler) getViolationContext(
+	policy *policyv1.Policy,
+	targetList []string,
+	policyAutomation *policyv1beta1.PolicyAutomation,
+) (policyv1beta1.ViolationContext, error) {
+	log.V(3).Info("Get the violation context from the root policy %s", policy.ObjectMeta.Name)
+
+	violationContext := policyv1beta1.ViolationContext{}
+	// 1) get the target cluster list
+	violationContext.TargetClusters = targetList
+	// 2) get the root policy name
+	violationContext.PolicyName = policy.ObjectMeta.Name
+	// 3) get the root policy namespace
+	violationContext.Namespace = policy.GetNamespace()
+	// 4) get the root policy hub cluster name
+	violationContext.HubCluster = policy.GetClusterName()
+
+	// 5) get the latest violation message from the root policy
+	statusDetails := policy.Status.Details
+	if len(statusDetails) > 0 && len(statusDetails[0].History) > 0 {
+		violationContext.ViolationMessage = statusDetails[0].History[0].Message
+	}
+
+	// 6) get the policy set of the root policy
+	plcPlacement := policy.Status.Placement
+	policySet := []string{}
+
+	for _, placement := range plcPlacement {
+		if placement.PolicySet != "" {
+			policySet = append(policySet, placement.PolicySet)
+		}
+	}
+
+	if len(policySet) > 0 {
+		violationContext.PolicySet = policySet
+	}
+
+	replicatedPlcList := &policyv1.PolicyList{}
+
+	err := r.List(
+		context.TODO(),
+		replicatedPlcList,
+		client.MatchingLabels(common.LabelsForRootPolicy(policy)),
+	)
+	if err != nil {
+		log.Error(err, "Failed to list the replicated policies")
+
+		return violationContext, err
+	}
+
+	if len(replicatedPlcList.Items) == 0 {
+		log.V(2).Info("The replicated policies can not be found.")
+	} else {
+		policyViolationContextLimit := policyAutomation.Spec.Automation.PolicyViolationContextLimit
+		if policyViolationContextLimit == 0 {
+			policyViolationContextLimit = 1000
+		}
+		targetListMap := getTargetListMap(targetList)
+		violationContext.PolicyViolationContext = make(map[string]policyv1.ReplicatedPolicyStatus)
```

Comment:
Optional: A small performance optimization.
```suggestion
		violationContext.PolicyViolationContext = make(map[string]policyv1.ReplicatedPolicyStatus, 0, len(replicatedPlcList.Items))
```